core/remotes/docker/resolver.go

```diff
@@ -654,9 +654,12 @@ func (r *request) retryRequest(ctx context.Context, responses []*http.Response)
 		}
 	case http.StatusRequestTimeout, http.StatusTooManyRequests:
 		return true, nil
+	case http.StatusInternalServerError, http.StatusBadGateway,
+		http.StatusServiceUnavailable, http.StatusGatewayTimeout:
+		// These 50X errors might be temporary so do retry
+		return true, nil
 	}
 
-	// TODO: Handle 50x errors accounting for attempt history
```

Comment:
Removing this comment - this adds retry for the 50X errors that are likely to be temporary. 
Not sure what could we do here to handle `accounting for attempt history` - retries are limited to 5 attempts already.
But happy to hear if there was a specific idea for this.
cc @dmcgowan

Reply:
I believe the idea was to look at the previous response and if you get the same 50x error then not retry again. If the first attempt is to a mirror which is timing out, trying 5 times can make the whole pull much slower. Could also be worth considering in this case whether this is the last registry endpoint which will be attempted.

Reply:
> I believe the idea was to look at the previous response and if you get the same 50x error then not retry again.
Hmm, I think that would make sense only for the `Gateway Timeout`. Others should in most cases probably fail early and next retries might actually work.

How about limiting the time the retries can take? For example for the 50X errors (except Gateway Timeout), we could limit the 1st retry to 5 seconds, 2nd retry to 4 seconds, etc.. 

>  If the first attempt is to a mirror which is timing out, trying 5 times can make the whole pull much slower
Right, I like the idea of limiting the retries for mirrors. However, I think we can't determine if the given host is a mirror without changing the `RegistryHosts` function to also include that information in `RegistryHost`?